### PR TITLE
Port libtsn to rust

### PR DESCRIPTION
**Pull request by MichelleJin12**
Add libtsn for rust and latency example

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,5 @@
 build/
+
+# rust
+target
+Cargo.lock
```
Comment (tribela) [inline .gitignore:5]: Cargo.lock should be in the repository

File: tsn/src/bin/latency.rs [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    sock: i32,
+    mut iface: String,
+    size: i32,
+    target: String,
+    count: i32,
+    precise: bool,
+    oneway: bool,
+) {
+    unsafe {
+        let pkt = libc::malloc(size as usize);
```
Comment (tribela) [inline tsn/src/bin/latency.rs:174]: Why not using bytearray and convert it at send?

File: tsn/src/bin/latency.rs [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+                    700 * 1000 * 1000 + (rand::thread_rng().gen_range(0..32767) as i64 % 10000000);
+                libc::nanosleep(&request, std::ptr::null_mut::<libc::timespec>());
+            }
+        }
+    }
+}
+
+fn main() {
+    unsafe {
+        let arguments = std::env::args();
```
Comment (tribela) [inline tsn/src/bin/latency.rs:346]: use [clap][] crate

[clap]: https://docs.rs/clap/2.33.0/clap/

File: tsn/src/bin/latency.rs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        let arguments = arguments::parse(arguments).unwrap();
+        let mut args: Arguments = Arguments {
+            ..Default::default()
+        };
+
+        let mode = arguments
+            .get::<String>("mode")
+            .expect("Need mode to run server or client");
+        if mode == "s" {
+            args = Arguments {
```
Comment (tribela) [inline tsn/src/bin/latency.rs:356]: Make default values for these unnecessary values

File: tsn/src/bin/latency.rs [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+                    .expect("Send packet at precise 0ns or not"),
+                oneway: arguments
+                    .get::<bool>("oneway")
+                    .expect("Check latency on receiver side"),
+            };
+        } else {
+            println!("Unknown mode");
+        }
+
+        SOCK = tsn::tsn_sock_open(
```
Comment (tribela) [inline tsn/src/bin/latency.rs:399]: ALLCAPS naming is for constant value

File: tsn/src/bin/latency.rs [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+            VLAN_PRI_PERF,
+            ETHERTYPE_PERF,
+        );
+
+        if SOCK <= 0 {
+            println!("socket create error");
+            panic!("last OS error: {:?}", Error::last_os_error());
+        }
+
+        libc::signal(libc::SIGINT, sigint as usize);
```
Comment (tribela) [inline tsn/src/bin/latency.rs:411]: Use standard rust's signal handler

File: tsn/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    vlanid: u32,
+}
+
+static mut SOCKETS: [TsnSocket; 20] = [TsnSocket {
+    fd: 0,
+    ifname: [0; 40],
+    vlanid: 0,
+}; 20];
+
+const CONTROL_SOCK_PATH: &str = "/var/run/tsn.sock\x00";
```
Comment (tribela) [inline tsn/src/lib.rs:19]: `String` can be automatically converted to `&str`

File: tsn/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+    if res < 0 {
+        libc::close(client_fd);
+        panic!("last OS error: {:?}", Error::last_os_error());
+    }
+}
+
+unsafe fn create_vlan(ifname: &str, vlanid: u32) {
+    let mut command: [libc::c_char; 128] = [0; 128];
+    libc::snprintf(
```
Comment (tribela) [inline tsn/src/lib.rs:74]: Why snprintf? Just use rust's string concat

**Review by tribela — CHANGES_REQUESTED** [review]
Converting every single line with libc crate is not porting

File: tsn/src/bin/latency.rs (hunk 5 above)
Comment (MichelleJin12) [reply tsn/src/bin/latency.rs:399]: Rust compiler warns that global variables don't have ALLCAPS names

File: tsn/src/lib.rs (hunk 7 above)
Comment (MichelleJin12) [reply tsn/src/lib.rs:19]: Compile errors about not providing a type for the constant.
```
error: missing type for `const` item
  --> src/lib.rs:21:7
   |
21 | const CONTROL_SOCK_PATH = "/var/run/tsn.sock\x00";
   |       ^^^^^^^^^^^^^^^^^ help: provide a type for the constant: `CONTROL_SOCK_PATH: &str`

error: could not compile `tsn` due to previous error
```

File: tsn/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    send_cmd(command)
+}
+
+fn delete_vlan(ifname: &str, vlanid: u32) -> Result<String, std::io::Error> {
+    let command = format!("delete {} {}\n", ifname, vlanid);
+    send_cmd(command)
+}
+
+#[allow(clippy::missing_safety_doc)]
+pub unsafe fn tsn_sock_open(
```
Comment (tribela) [inline tsn/src/lib.rs:36]: Could you remove `unsafe` on functions?

File: tsn/src/bin/latency.rs [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+use std::{thread, time::Duration};
+
+extern crate hex;
+extern crate ifstructs;
+extern crate socket as soc;
+
+const VLAN_ID_PERF: u32 = 10;
+const VLAN_PRI_PERF: u32 = 3;
+const ETHERTYPE_PERF: u32 = 0x1337;
+static mut RUNNING: i32 = 1;
```
Comment (tribela) [inline tsn/src/bin/latency.rs:23]: Use atomic boolean, then remove unsafe blocks related to this

File: tsn/src/bin/latency.rs [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+            println!("Interrrupted");
+            RUNNING.fetch_and(false, Ordering::Relaxed);
+            unsafe {
+                tsn::tsn_sock_close(&mut SOCK);
+            }
+            std::process::exit(1);
+        }
+    });
+
+    if mode == "s" {
```
Comment (tribela) [inline tsn/src/bin/latency.rs:434]: use `match`

**Review by tribela — APPROVED** [review]

**Pull request merged**